File: (6 kyu) Help the bookseller/(6 kyu) Help the bookseller.js
// #1 Use forEach
// function stockList(listOfArt, listOfCat) {
//   if (!listOfArt.length || !listOfCat.length) {
//     return '';
//   }
//   const books = {};
//   const result = [];
//   listOfArt.forEach((book) => {
//     const [title, num] = book.split` `;
//     books[title[0]] = books[title[0]] ? (books[title[0]] += +num) : +num;
//   });
//   listOfCat.forEach((letter) => {
//     const num = books[letter] || 0;
//     result.push(`(${letter} : ${num})`);
//   });
//   return result.join` - `;
// }

// #2 Optimazed forEach
// function stockList(listOfArt, listOfCat) {
//   if (!listOfArt.length || !listOfCat.length) {
//     return '';
//   }
//   const books = {};
//   listOfArt.forEach((book) => {
//     const title = book[0];
//     books[title] = (books[title] | 0) + +book.split` `[1];
//   });
//   return listOfCat.map((letter) => {
//     return `(${letter} : ${books[letter] | 0})`;
//   }).join` - `;
// }

// #3 Use reduce
function stockList(listOfArt, listOfCat) {
  if (!listOfArt.length || !listOfCat.length) {
    return '';
  }
  return listOfCat
    .map((book) => {
      const sum = listOfArt.reduce(
        (acc, title) => acc + (title[0] === book ? +title.split(' ')[1] : 0),
        0,
      );
      return `(${book} : ${sum})`;
    })
    .join(' - ');
}

File: (6 kyu) Help the bookseller/(6 kyu) Help the bookseller.rs
use std::collections::HashMap;

fn stock_list(list_art: Vec<&str>, list_cat: Vec<&str>) -> String {
    if list_art.len() == 0 || list_cat.len() == 0 {
        return String::from("");
    }
    let mut books: HashMap<String, i32> = HashMap::new();
    let mut result = Vec::new();
    
    for book in list_art {
        let title_num: Vec<&str> = book.split_whitespace().collect();
        let title: String = title_num[0][0..1].to_string();
        let num: i32 = title_num[1].parse().unwrap();

        *books.entry(title).or_insert(0) += num;
      }
      
    for letter in list_cat {
        let num = match books.get(letter) {
            Some(n) => *n,
            None => 0
        };
        result.push(format!("({} : {})", letter, num))
    }
    result.join(" - ")
}
